frame/staking-rewards/src/lib.rs

```diff
@@ -599,8 +626,118 @@ pub mod pallet {
+							.saturating_mul(reward.reward_rate.amount.into())
+							.saturating_add(reward.total_rewards.into());
+
+						let new_total_rewards = if new_total_rewards <= reward.max_rewards.into() {
+							new_total_rewards.into()
+						} else {
+							// saturate at max_rewards, but emit an error first
+							// only emit if the previous total_rewards wasn't at the max_rewards
+							// (i.e. the event would have been emitted already)
+							if reward.total_rewards < reward.max_rewards {
+								Self::deposit_event(Event::<T>::MaxRewardsAccumulated {
+									pool_id,
+									asset_id: reward.asset_id,
+								});
+							}
+							reward.max_rewards
+						};
+
+						Reward {
+							total_rewards: new_total_rewards,
+							last_updated_timestamp: reward
+								.last_updated_timestamp
+								.add(periods_surpassed.mul(reward_rate_period_seconds)),
+							..reward
+						}
+					}
+				},
+				Err(_) => {
+					Self::deposit_event(Event::<T>::RewardAccumulationError {
+						pool_id,
+						asset_id: reward.asset_id,
+					});
+					reward
+				},
+			}
+		}
+
+		pub(crate) fn acumulate_rewards_hook() -> Weight {
+			let now = T::UnixTime::now().as_secs();
+			let unix_time_now_weight = T::WeightInfo::unix_time_now();
+
+			let updated_pools = RewardPools::<T>::iter()
+				.into_iter()
+				.map(|(pool_id, reward_pool)| {
+					let updated_rewards = reward_pool
+						.rewards
+						.into_iter()
+						.map(|(asset_id, reward)| {
+							(
+								asset_id,
+								Self::reward_acumulation_hook_reward_update_calculation(
+									pool_id, reward, now,
+								),
+							)
+						})
+						.try_collect()
+						// SAFETY(benluelo): No elements were added to the BTreeMap; the only
+						// operation was `.map()`. This expect call will be unnecessary once this PR
+						// is merged and we update to whatever version it's included in:
+						// https://github.com/paritytech/substrate/pull/11869
+						.expect("no elements were added; qed;");
+
+					(pool_id, RewardPool { rewards: updated_rewards, ..reward_pool })
+				})
+				// NOTE(benluelo): As per these docs on `StorageMap::iter`:
+				// https://github.com/paritytech/substrate/blob/cac91f59b9e3fb8fd59842c023f87b4206931993/frame/support/src/storage/types/map.rs,
+				// "If you alter the map while doing this, you'll get undefined results."
+				// hence the double iteration.
+				.collect::<Vec<_>>();
+
+			let mut total_weight = unix_time_now_weight;
+
+			for (pool_id, reward_pool) in updated_pools {
+				// 128 bit platforms don't exist as of writing this so this usize -> u64 cast should
+				// be ok
+				let number_of_rewards_in_pool = reward_pool.rewards.len() as u64;
+
+				RewardPools::<T>::insert(pool_id, reward_pool);
+
+				total_weight += (number_of_rewards_in_pool * T::WeightInfo::reward_acumulation_hook_reward_update_calculation()) +
+						// REVIEW(benluelo): I'm assuming that `StorageMap::iter` does one read per item, I could be wrong though
```

Comment:
I looked pallets of substrate some time ago and came to the same conclusion.

Reply:
Coo, good to know. Updated the comment :+1: